src/mlpack/tests/main_tests/range_search_test.cpp

```diff
@@ -0,0 +1,292 @@
+    SetInputParam("min", min_v);
+    SetInputParam("max", max_v);
+    SetInputParam("distances_file", distance_file);
+    SetInputParam("neighbors_file", neighbors_file);
+
+    mlpackMain();
+
+    RSModel* output_model2 = std::move(CLI::GetParam<RSModel*>("output_model"));
+    output_model2->Search(r, neighbors_temp, distances_temp);
+
+    CheckMatrices(neighbors, neighbors_temp, 1e-5);
+    CheckMatrices(distances, distances_temp);
+  }
+  cout<<"Leaf value Test Passed"<<endl<<endl;
+}
+
+//All trees should give the same results for fixed input parameters
+BOOST_AUTO_TEST_CASE(TreeTypeTesting)
+{
+  cout<<"Tree Type Testing"<<endl;
+  std::string distance_file = "distances.csv";
+  std::string neighbors_file = "neighbors.csv";
+  double min_v = 0, max_v = 3;
+  arma::mat query_data, input_data;
+  vector<vector<size_t>> neighbors, neighbors_temp;
+  vector<vector<double>> distances, distances_temp;
+  std::vector<string> trees = {"kd", "cover", "r", "r-star", "ball", "x",
+                               "hilbert-r", "r-plus", "r-plus-plus", "vp","rp",
+                               "max-rp", "ub", "oct"};
+
+
+  if (!data::Load("iris.csv", input_data))
+    BOOST_FAIL("Unable to load dataset iris.csv!");
+  if (!data::Load("iris_test.csv", query_data))
+    BOOST_FAIL("Unable to load dataset iris_test.csv!");
+  math::Range r(min_v, max_v);
+  //Define Base with kd Tree
+  SetInputParam("tree_type", trees[0]);
+  SetInputParam("min", min_v);
+  SetInputParam("max", max_v);
+  SetInputParam("distances_file", distance_file);
+  SetInputParam("neighbors_file", neighbors_file);
+  SetInputParam("reference", std::move(input_data));
+  SetInputParam("query", query_data);
+
+  mlpackMain();
+
+  cout<<"  Created Base value with kd tree"<<endl;
+  CLI::GetParam<RSModel*>("output_model")->Search(std::move(query_data), r,
+                                                  neighbors, distances);
+
+  for (size_t i = 1;i < trees.size() ; i++)
+  {
+    bindings::tests::CleanMemory();
+    cout<<"  Testing for Tree type :"<<trees[i]<<endl;
+    if (!data::Load("iris.csv", input_data))
+      BOOST_FAIL("Unable to load dataset iris.csv!");
+    if (!data::Load("iris_test.csv", query_data))
+      BOOST_FAIL("Unable to load dataset iris_test.csv!");
+    math::Range r(min_v, max_v);
+
+    SetInputParam("min", min_v);
+    SetInputParam("max", max_v);
+    SetInputParam("distances_file", distance_file);
+    SetInputParam("neighbors_file", neighbors_file);
+    SetInputParam("reference", std::move(input_data));
+    SetInputParam("query", query_data);
+    SetInputParam("tree_type", trees[i]);
+
+    mlpackMain();
+
+    CLI::GetParam<RSModel*>("output_model")->Search(std::move(query_data), r,
+                                                neighbors_temp, distances_temp);
+    CheckMatrices(neighbors, neighbors_temp);
+    CheckMatrices(distances, distances_temp);
+    cout<<"    Successful"<<endl;
+  }
+}
```

Comment:
We could also check here that the trained models are different after serialization, just to ensure that the `tree_type` parameter is not ignored.